chronicle-queue/src/test/scala/akka/stream/alpakka/chroniclequeue/ChronicleQueueAtLeastOnceSpec.scala

```diff
@@ -0,0 +1,355 @@
+    count shouldBe elementCount
+    totalProcessed shouldBe elementCount
+    t1 should be > t2 // Give 6 seconds difference. In fact, it should be closer to 9 seconds.
+    clean()
+  }
+
+  it should "recover from unexpected stream shutdown" in {
+    implicit val util = new StreamSpecUtil[T, Event[T]]
+    import util._
+
+    val mat = ActorMaterializer()
+    val pBufferInCount = new AtomicInteger(0)
+    val commitCount = new AtomicInteger(0)
+    val finishedGenerating = Promise[Done]
+    val counter = new AtomicInteger(0)
+
+    def fireFinished() =
+      Flow[T]
+        .map { e =>
+          if (counter.incrementAndGet() == failTestAt) finishedGenerating.success(Done)
+          e
+        }
+        .toMat(Sink.ignore)(Keep.right)
+
+    val shutdownF = finishedGenerating.future.map { d =>
+      mat.shutdown(); d
+    }
+
+    val graph = RunnableGraph.fromGraph(GraphDSL.create(Sink.ignore) { implicit builder => sink =>
+      import GraphDSL.Implicits._
+      val buffer = ChronicleQueueAtLeastOnce[T](config)
+        .withOnPushCallback(() => pBufferInCount.incrementAndGet())
+        .withOnCommitCallback(() => commitCount.incrementAndGet())
+      val commit = buffer.commit[T] // makes a dummy flow if autocommit is set to false
+      val bc = builder.add(Broadcast[T](2))
+
+      in ~> transform ~> bc ~> buffer.async ~> throttle ~> commit ~> sink
+      bc ~> fireFinished()
+
+      ClosedShape
+    })
+    val sinkF = graph.run()(mat)
+    Await.result(shutdownF, awaitMax)
+    Await.result(sinkF.failed, awaitMax) shouldBe an[AbruptTerminationException]
+
+    val restartFrom = pBufferInCount.incrementAndGet()
+    println(s"Restart from count $restartFrom")
+
+    resumeGraphAndDoAssertion(commitCount.get, restartFrom)
+    clean()
+  }
+
+  it should "recover from downstream failure" in {
+    implicit val util = new StreamSpecUtil[T, Event[T]]
+    import util._
+
+    val mat = ActorMaterializer()
+    val outCount = new AtomicInteger(0)
+    val injectCounter = new AtomicInteger(0)
+    val inCounter = new AtomicInteger(0)
+
+    val injectError = Flow[Event[T]].map { n =>
+      val count = injectCounter.incrementAndGet()
+      if (count == failTestAt) throw new NumberFormatException("This is a fake exception")
+      else n
+    }
+
+    val graph = RunnableGraph.fromGraph(GraphDSL.create(Sink.ignore) { implicit builder => sink =>
+      import GraphDSL.Implicits._
+      val buffer = ChronicleQueueAtLeastOnce[T](config)
+        .withOnPushCallback(() => inCounter.incrementAndGet())
+        .withOnCommitCallback(() => outCount.incrementAndGet())
+      val commit = buffer.commit[T] // makes a dummy flow if autocommit is set to false
+      in ~> transform ~> buffer.async ~> throttle ~> injectError ~> commit ~> sink
+      ClosedShape
+    })
+    val sinkF = graph.run()(mat)
+    Await.result(sinkF.failed, awaitMax) shouldBe an[NumberFormatException]
```

Comment:
Shouldn't have to use `Await`. Consider ScalaTest's future handling functionality.

Reply:
changed to `futureValue` all around